fvm/src/kernel/default.rs

```diff
@@ -441,6 +456,29 @@ where
     }
 }
 
+impl<C> EnvOps for DefaultKernel<C>
+where
+    C: CallManager,
+{
+    /// current tipset timestamp
+    fn tipset_timestamp(&self) -> u64 {
+        self.call_manager.chain_context().timestamp
+    }
+
+    /// epoch tipset cid
+    fn tipset_cid(&self, epoch: i64) -> Result<Option<Cid>> {
+        if epoch < 0 {
+            return Err(syscall_error!(IllegalArgument; "epoch is negative").into());
+        }
+        let tipsets = &self.call_manager.chain_context().tipsets;
+        if (epoch as usize) < tipsets.len() {
+            return Ok(Some(tipsets[epoch as usize]));
+        }
```

Comment:
For extra safety and formality, we should _also_ validate that the specific lookback is within the admitted range (256 for Ethereum, 900 if we want to follow Filecoin's finality).

Reply:
this should probably happen at call site? We don't know if it is an eth or fil invocation here, so we merely check against the length of the vector.

Reply:
The FVM should place a hard limit. I would vote for 900 epochs because it matches our finality. The EVM runtime should validate that the argument is 256 max.

Reply:
ok, 900 it is -- so we get an error (as opposed to a None) if we pass more than that.